### PR TITLE
Add etag and cache control

### PR DESCRIPTION
**Pull request by bvallelunga**

### CODE REVIEW

File: lib/piler.coffee [hunk 1]
```diff
@@ -18,6 +18,10 @@ toGlobals = (globals) ->
     code += "__SET(#{ JSON.stringify nsString }, #{ OB.stringify v });\n"
   code
 
+hasher = (string)->
+  sum = crypto.createHash('sha1')
```
Comment (Piccirello) [inline lib/piler.coffee]: In general let's default to using `sha256` unless we have a good reason not to. sha1 is more vulnerable to hash collisions.

**Comment by bvallelunga** [comment]
> I've been doing some research and want to modify my proposal. I think we can achieve the best cache rates by continuing to use the `v` parameter, but modifying it to serve the asset's hash (like the ETag). This removes the need for the ETag, though there's no harm in leaving it in.
> 
> Once we have this, we're now serving our assets behind immutable urls- if the content ever changes, the url changes. This allows us to use a crazy `Cache-Control` header like `"public, immutable, must-revalidate, max-age=31536000"`, which translates to:
> 
> 1. cache everything in the browser and in Cloudflare for 1 year (`public` and `max-age`)
> 2. never try to verify if an asset is still valid before its expiration; just trust it (`immutable`)**
> 3. once the year is up, immediately stop using the cached version (`must-revalidate`)
> 
> I've previously implemented this approach at BlackRock. It's the same approach used by Facebook, Cloudflare, GitHub (for some assets), etc.
> 
> **Two things I want to confirm:**
> 
> 1. piler is only serving static assets. None of the files are dynamically generated/modified before being served, and none are unique per-user.
> 2. the file hash is cached by piler and not recomputed every time an asset is requested (for performance)
> 
> Happy to discuss all this.
> 
> ** immutable is not implemented in every browser yet. these browsers may still request the content but will rely on the etag to determine if a response must be redownloaded

I dig this. Happy to make the change.

In response:

1. They are static assets that are generated at boot time. They are user specific.
2. Currently it is setup to hash on every request. I can look into if I can cache it.

File: lib/piler.coffee [hunk 2]
```diff
@@ -329,17 +334,17 @@ class PileManager
         res.send "Cannot find pile #{ asset.name }", 404
         return
 
-      res.setHeader 'Cache-Control', 'public, max-age=0'
-
-      # TODO: set cache headers to forever
       if asset.min
+        res.setHeader("Etag", hasher(pile.rawPile))
```
[the remark this answers is not in the document]
Comment (bvallelunga) [reply lib/piler.coffee]: yes that is correct

File: lib/piler.coffee [hunk 3]
```diff
@@ -18,6 +18,10 @@ toGlobals = (globals) ->
     code += "__SET(#{ JSON.stringify nsString }, #{ OB.stringify v });\n"
   code
 
+hasher = (string)->
+  sum = crypto.createHash('sha1')
+  sum.update string
+  return sum.digest('hex').substring(10, 0)
```
[the remark this answers is not in the document]
Comment (bvallelunga) [reply lib/piler.coffee]: im game.

[the remark this answers is not in the document]
Comment (bvallelunga) [reply lib/piler.coffee]: honestly not sure, I copied it from another part of the code. Will update to your recommended length anyways.

File: lib/piler.coffee (hunk 1 above)
Comment (bvallelunga) [reply lib/piler.coffee]: im game, nice catch!

**Comment by Piccirello** [comment]
One edit to my earlier comment: I don't believe it completely removes the need for an ETag. The ETag may still be helpful for preventing Cloudflare from hitting our Origin.

File: lib/piler.coffee [hunk 4]
```diff
@@ -143,7 +142,7 @@ class BasePile
     (codeOb for codeOb in @code when codeOb.filePath is id)[0]
 
   _computeHash: ->
-    sum = crypto.createHash('sha1')
+    sum = crypto.createHash('sha256')
```
Comment (Piccirello) [inline lib/piler.coffee]: There's one other instance of `sha1` being used- can we upgrade that too in a separate commit?

File: lib/piler.coffee [hunk 5]
```diff
@@ -129,7 +128,7 @@ class BasePile
     sources = ([u] for u in @urls)
 
     if @production
-      sources.push ["#{ @urlRoot }min/#{ @pileHash }/#{ @name }.#{ @ext }"]
```
Comment (Piccirello) [inline lib/piler.coffee]: How is it that the hash was already previously in the url?

File: lib/piler.coffee [hunk 6]
```diff
@@ -129,7 +128,7 @@ class BasePile
...
+      sources.push ["#{ @urlRoot }min/#{ @name }.#{ @ext }?v=#{@pileHash}"]
```
Comment (Piccirello) [inline lib/piler.coffee]: I wonder if we can get in trouble here or with the ETag if the hash is super long. Theoretically I suppose the url could exceed the max length (~2000 characters), or the ETag could cause the http header to exceed the max header size (4K-8K). Don't know how practical this concern is- meaning how large a file (+ potentially other properties) has to be for the hash to get so big.

File: lib/piler.coffee (hunk 5 above)
Comment (bvallelunga) [reply lib/piler.coffee]: Yea i guess it was, I didn't realize until when I made the last commit.

File: lib/piler.coffee (hunk 6 above)
Comment (bvallelunga) [reply lib/piler.coffee]: ill cap it at 25 characters.

File: lib/piler.coffee (hunk 5 above)
Comment (Piccirello) [reply lib/piler.coffee]: Oh snap, they totally are: `https://doppler.com/css/min/5ba2318f73a7dab33c123fa8e269c868e79ba6b5/global.css?v=xohbPh1ebP`. I didn't even realize. Can we go back to the original logic? I personally prefer the current approach with the hash in the url.

For these we might as well completely remove the query string. But is there an easy way to do that without also removing the query string from the `external.txt` files?

File: lib/piler.coffee (hunk 6 above)
Comment (Piccirello) [reply lib/piler.coffee]: Based on our previous usage of the full hash I think it's safe to use the whole thing.

File: lib/piler.coffee (hunk 5 above)
Comment (bvallelunga) [reply lib/piler.coffee]: I don't think so, its either all or nothing. I prefer the hash in the url as well. So final game plan is to add the hash to the url, use etags, and remove the query parameter right?

File: lib/piler.coffee (hunk 4 above)
Comment (bvallelunga) [reply lib/piler.coffee]: Not sure why github isn't saying this is outdated, but it has been changed.

File: lib/piler.coffee (hunk 5 above)
Comment (Piccirello) [reply lib/piler.coffee]: As discussed offline we still need the query parameter for local assets loaded via `external.txt`, but we'll accomplish that in server code.

File: lib/piler.coffee [hunk 7]
```diff
@@ -316,7 +314,8 @@ class PileManager
       if not _.startsWith req.url, @settings.urlRoot
         return next()
 
-      res.setHeader "Content-type", @contentType
+      res.setHeader("Content-type", @contentType)
+      res.setHeader('Cache-Control', 'public, immutable, must-revalidate, max-age=31536000')
```
Comment (Piccirello) [inline lib/piler.coffee]: Will this also get set for assets added via `piler.addFile` or `piler.addUrl`? Or will we have to set `Cache-Control` manually for these?

Comment (bvallelunga) [reply lib/piler.coffee]: it will set it for `addFile` but not `addUrl`. We serve bower content through the `express.static` in `dashboard/index.coffee`

**Pull request merged**